File: app/src/main/java/com/olehmesh/asinfo2/WidgetAdapter.java
package com.olehmesh.asinfo2;

import java.util.ArrayList;
import android.annotation.TargetApi;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.BatteryManager;
import android.os.Build;
import android.widget.RemoteViews;
import android.widget.RemoteViewsService.RemoteViewsFactory;

public class WidgetAdapter implements RemoteViewsFactory {

    ArrayList<String> list;
    Context context;

    WidgetAdapter(Context ctx, Intent intent) {
        context = ctx;
    }

    @Override
    public void onCreate() {
        list = new ArrayList<>();
    }

    @Override
    public int getCount() {
        return list.size();
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public RemoteViews getLoadingView() {
        return null;
    }

    @Override
    public RemoteViews getViewAt(int position) {
        RemoteViews remoteViews = new RemoteViews(context.getPackageName(),
                R.layout.list_item);
        remoteViews.setTextViewText(R.id.itemText, list.get(position));

        Intent clickIntent = new Intent();
        clickIntent.putExtra(MyProvider.ITEM_POSITION, position);
        remoteViews.setOnClickFillInIntent(R.id.itemText, clickIntent);

        return remoteViews;
    }

    @Override
    public int getViewTypeCount() {
        return 1;
    }

    @Override
    public boolean hasStableIds() {
        return true;
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    @Override
    public void onDataSetChanged() {

        list.clear();

        BatteryManager bm = (BatteryManager) context.getSystemService(Context.BATTERY_SERVICE);
        Intent intent = context.registerReceiver(null, new IntentFilter(Intent.ACTION_BATTERY_CHANGED));

        int mHealth = intent.getIntExtra(BatteryManager.EXTRA_HEALTH, 0);

        list.add("Battery health: " + convertHealth(mHealth));

        int mTemp = intent.getIntExtra(BatteryManager.EXTRA_TEMPERATURE, 0);
        float temp = mTemp / 10.0f;

        list.add("Temperature: " + temp + " °C");

        int mVoltage = intent.getIntExtra(BatteryManager.EXTRA_VOLTAGE, 0);
        float tempTurn = mVoltage / 1000.0f;
        list.add("Voltage: " + tempTurn + " V");

        int current = bm.getIntProperty(BatteryManager.BATTERY_PROPERTY_CURRENT_NOW);
        float currTurn = current / 1000.0f;
        list.add("Current: " + currTurn + " mA");

    }

    private String convertHealth(int health){
        String result;
        switch(health){
            case BatteryManager.BATTERY_HEALTH_COLD:
                result = "Cold";
                break;
            case BatteryManager.BATTERY_HEALTH_DEAD:
                result = "Dead";
                break;
            case BatteryManager.BATTERY_HEALTH_GOOD:
                result = "Good";
                break;
            case BatteryManager.BATTERY_HEALTH_OVERHEAT:
                result = "Overheat";
                break;
            case BatteryManager.BATTERY_HEALTH_OVER_VOLTAGE:
                result = "Over voltage";
                break;
            case BatteryManager.BATTERY_HEALTH_UNKNOWN:
                result = "Unknown";
                break;
            case BatteryManager.BATTERY_HEALTH_UNSPECIFIED_FAILURE:
                result = "Unspecified failure";
                break;
            default:
                result = "Unknown";
        }

        return result;
    }

    @Override
    public void onDestroy() {

    }

}

File: app/src/main/java/com/olehmesh/asinfo2/WidgetAdapter.kt
package com.olehmesh.asinfo2

import android.annotation.TargetApi
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.BatteryManager
import android.os.Build
import android.widget.RemoteViews
import android.widget.RemoteViewsService.RemoteViewsFactory
import java.util.*

class WidgetAdapter internal constructor(var context: Context, intent: Intent?) :
    RemoteViewsFactory {
    var list: ArrayList<String>? = null
    override fun onCreate() {
        list = ArrayList()
    }

    override fun getCount(): Int {
        return list!!.size
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getLoadingView(): RemoteViews? {
        return null
    }

    override fun getViewAt(position: Int): RemoteViews {
        val remoteViews = RemoteViews(
            context.packageName,
            R.layout.list_item
        )
        remoteViews.setTextViewText(R.id.itemText, list!![position])
        val clickIntent = Intent()
        clickIntent.putExtra(MyProvider.ITEM_POSITION, position)
        remoteViews.setOnClickFillInIntent(R.id.itemText, clickIntent)
        return remoteViews
    }

    override fun getViewTypeCount(): Int {
        return 1
    }

    override fun hasStableIds(): Boolean {
        return true
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    override fun onDataSetChanged() {
        list!!.clear()
        val bm = context.getSystemService(Context.BATTERY_SERVICE) as BatteryManager
        val intent = context.registerReceiver(null, IntentFilter(Intent.ACTION_BATTERY_CHANGED))
        val mHealth = intent!!.getIntExtra(BatteryManager.EXTRA_HEALTH, 0)
        list!!.add("Battery health: " + convertHealth(mHealth))
        val mTemp = intent.getIntExtra(BatteryManager.EXTRA_TEMPERATURE, 0)
        val temp = mTemp / 10.0f
        list!!.add("Temperature: $temp °C")
        val mVoltage = intent.getIntExtra(BatteryManager.EXTRA_VOLTAGE, 0)
        val tempTurn = mVoltage / 1000.0f
        list!!.add("Voltage: $tempTurn V")
        val current = bm.getIntProperty(BatteryManager.BATTERY_PROPERTY_CURRENT_NOW)
        val currTurn = current / 1000.0f
        list!!.add("Current: $currTurn mA")
    }

    private fun convertHealth(health: Int): String {
        val result: String
        result = when (health) {
            BatteryManager.BATTERY_HEALTH_COLD -> "Cold"
            BatteryManager.BATTERY_HEALTH_DEAD -> "Dead"
            BatteryManager.BATTERY_HEALTH_GOOD -> "Good"
            BatteryManager.BATTERY_HEALTH_OVERHEAT -> "Overheat"
            BatteryManager.BATTERY_HEALTH_OVER_VOLTAGE -> "Over voltage"
            BatteryManager.BATTERY_HEALTH_UNKNOWN -> "Unknown"
            BatteryManager.BATTERY_HEALTH_UNSPECIFIED_FAILURE -> "Unspecified failure"
            else -> "Unknown"
        }
        return result
    }

    override fun onDestroy() {}

}